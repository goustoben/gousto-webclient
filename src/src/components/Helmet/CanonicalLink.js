import React, { PropTypes } from 'react'
import Helmet from 'react-helmet'

const CanonicalLink = ({ href }) => (
	<Helmet
	  link={[
	    {
	      rel: 'canonical',
	      href,
	    },
	  ]}
	/>
)

CanonicalLink.propTypes = {
  href: PropTypes.string.isRequired,
}

export default CanonicalLink
