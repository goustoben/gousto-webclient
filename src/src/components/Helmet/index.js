export { default as Component } from './GoustoHelmet'
