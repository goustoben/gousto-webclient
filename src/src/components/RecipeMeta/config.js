export const perPortionPrice = 2.98
export const defaultMetaImageLink = "https://production-media.gousto.co.uk/cms/promo-image/5c38625bdc0e82003eaf3737-1125x750.jpg"