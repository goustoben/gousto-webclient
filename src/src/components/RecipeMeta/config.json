{
  "price": 2.98,
  "defaultImageLink": "https://production-media.gousto.co.uk/cms/promo-image/5c38625bdc0e82003eaf3737-1125x750.jpg"
}
