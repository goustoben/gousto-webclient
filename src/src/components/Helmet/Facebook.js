import PropTypes from 'prop-types'
import React from 'react'
import Helmet from 'react-helmet'

const Facebook = ({ admins, appID, domainVerification }) => {
  const metaData = [
    {
      property: 'fb:app_id',
      content: appID,
    },
    {
      name: 'facebook-domain-verification',
      content: domainVerification,
    }
  ]

  if (admins && admins.length) {
    admins.forEach(admin => metaData.push({
      property: 'fb:admins',
      content: admin,
    }))
  }

  return <Helmet meta={metaData} />
}

Facebook.propTypes = {
  admins: PropTypes.array.isRequired,
  appID: PropTypes.string.isRequired,
  domainVerification: PropTypes.string.isRequired,
}

export default Facebook
