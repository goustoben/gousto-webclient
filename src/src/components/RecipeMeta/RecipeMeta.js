import React from 'react'
import Immutable from 'immutable'
import Helmet from 'react-helmet'
import PropTypes from 'prop-types'

import {
  getMetaImageLink,
  getPortionPrice,
  getRecipeDetailURL,
} from './utils'

const RecipeMeta = ({ recipe }) => (
  (recipe) ? (
    <Helmet
      meta={[
        {
          name: 'og:type',
          content: 'product',
        },
        {
          name: 'og:title',
          content: recipe.get('title'),
        },
        {
          name: 'og:description',
          content: recipe.get('description'),
        },
        {
          name: 'og:url',
          content: getRecipeDetailURL(recipe),
        },
        {
          name: 'og:image',
          content: getMetaImageLink(recipe),
        },
        {
          name: 'product:brand',
          content: recipe.getIn(['range', 'name'], ''),
        },
        {
          name: 'product:condition',
          content: 'new',
        },
        {
          name: 'product:price:amount',
          content: getPortionPrice(recipe),
        },
        {
          name: 'product:price:currency',
          content: 'GBP',
        },
        {
          name: 'product:retailer_item_id',
          content: recipe.get('id'),
        },
      ]}
    />
  ) : null
)

RecipeMeta.propTypes = {
  recipe: PropTypes.instanceOf(Immutable.Map),
}
