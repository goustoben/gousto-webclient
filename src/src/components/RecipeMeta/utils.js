import Immutable from 'immutable'

import { getSurcharge } from 'utils/recipe'

import { defaultMetaImageLink, perPortionPrice } from './config.js'

export const getPortionPrice = (recipe) => {
  const surcharge = getSurcharge(recipe.get('meals'), 4)

  return surcharge ? (perPortionPrice + surcharge / 4).toFixed(2) : perPortionPrice
}

export const getMetaImageLink = (recipe) => {
  const media = recipe.getIn(['media', 'images', 0, 'urls'], Immutable.List([]))
  const image = media.find(imageProps => imageProps.get('width') == 700)

  return image ? image.get('src') : defaultMetaImageLink
}

export const getRecipeDetailURL = (recipe) => (
  `https://www.gousto.co.uk/menu?recipeDetailId=${recipe.get('id')}`
)
