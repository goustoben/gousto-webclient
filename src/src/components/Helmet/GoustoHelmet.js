import PropTypes from 'prop-types'
import React from 'react'
import templateConfig from 'config/template'
import globals from 'config/globals'
import client from 'config/routes'
import CanonicalLink from './CanonicalLink'
import Facebook from './Facebook'
import OpenGraph from './OpenGraph'
import SiteVerification from './SiteVerification'
import Twitter from './Twitter'

export const generateHref = url => {
  const hostname = `${globals.env === 'production' ? 'www.' : `${globals.env}-frontend.`}${globals.domain}`
  const resultURL = url === client.menu2 ? client.menu : url

  return `${globals.protocol}://${hostname}${resultURL || '/'}`
}

const propTypes = {
  requestUrl: PropTypes.string.isRequired,
  scripts: PropTypes.shape({
    other: PropTypes.bool,
  }),
}

const defaultProps = {
  scripts: {
    other: true,
  }
}

const GoustoHelmet = ({ requestUrl, scripts: { other } }) => (
  other ? (
    <span>
      <CanonicalLink href={generateHref(requestUrl)} />
      <Facebook admins={templateConfig.head.fbAdmins} appID={templateConfig.head.fbAppID} />
      <OpenGraph href={generateHref(requestUrl)} />
      <Twitter href={generateHref(requestUrl)} />
      <SiteVerification />
    </span>
  ) : null
)

GoustoHelmet.propTypes = propTypes
GoustoHelmet.defaultProps = defaultProps

export default GoustoHelmet
