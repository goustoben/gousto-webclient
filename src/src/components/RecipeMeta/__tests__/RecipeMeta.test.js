import React from 'react'
import { shallow } from 'enzyme'
import Immutable from 'immutable'
import Helmet from 'react-helmet'

import { RecipeMeta } from 'RecipeMeta'

describe('RecipeMeta', () => {
  const recipe = Immutable.fromJS({
    id: '12345',
    title: 'A Recipe Title',
    description: 'The best you\'ve ever tasted.',
    media: {
      images: [{
        urls: [{
          src: 'protocol://imageurl.com/image.jpg',
          width: 700,
        }],
      }],
    },
    taxonomy: [
      {},
      {
        id: 2,
        name: 'Food Brands',
        slug: 'food-brands',
        tags: [
          {
            id: '9',
            name: 'Great Food',
            properties: {
              ribbon_color: '#000000',
              border_color: '#222222',
              text_color: '#FFFFFF',
            },
            slug: 'great-food',
          }
        ]
      }
    ]
  })

  it('should render helmet component if recipe exists', () => {
    const wrapper = shallow(<RecipeMeta recipe={recipe} />)

    expect(wrapper.find(Helmet).length).toEqual(1)
  })

  it('should render nothing if recipe does not exist', () => {
    const wrapper = shallow(<RecipeMeta recipe={null} />)

    expect(wrapper.find(Helmet).length).toEqual(0)
  })

  test('should render meta tags', () => {
    const wrapper = shallow(<RecipeMeta recipe={recipe} />)
    const meta = wrapper.find(Helmet).first().prop('meta')

    expect(meta).toMatchSnapshot()
  })
})
