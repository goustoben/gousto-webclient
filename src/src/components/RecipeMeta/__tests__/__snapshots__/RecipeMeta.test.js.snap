// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeMeta should render meta tags 1`] = `
Array [
  Object {
    "content": "product",
    "property": "og:type",
  },
  Object {
    "content": "A Recipe Title",
    "property": "og:title",
  },
  Object {
    "content": "The best you've ever tasted.",
    "property": "og:description",
  },
  Object {
    "content": "https://www.gousto.co.uk/menu?recipeDetailId=12345",
    "property": "og:url",
  },
  Object {
    "content": "protocol://imageurl.com/image.jpg",
    "property": "og:image",
  },
  Object {
    "content": "Great Food",
    "property": "product:brand",
  },
  Object {
    "content": "new",
    "property": "product:condition",
  },
  Object {
    "content": 2.98,
    "property": "product:price:amount",
  },
  Object {
    "content": "GBP",
    "property": "product:price:currency",
  },
  Object {
    "content": "12345",
    "property": "product:retailer_item_id",
  },
  Object {
    "content": "available for order",
    "property": "product:availability",
  },
]
`;
